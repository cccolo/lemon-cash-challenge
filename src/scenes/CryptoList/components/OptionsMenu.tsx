import React from 'react';
import {
  Animated,
  StyleSheet,
  Switch,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import {SearchIcon} from '../../../assets/images';
import {InputSearch} from './InputSearch';

type Props = {
  onToggleMenu: () => void;
  onAction: (type: 'SEARCH' | 'FAVORITES', value: string | boolean) => void;
};

export const OptionsMenu: React.FC<Props> = ({
  onToggleMenu,
  onAction,
}: Props) => {
  const [expanded, setExpanded] = React.useState(false);
  const [isEnabled, setIsEnabled] = React.useState(false);
  const animation = React.useRef(new Animated.Value(0)).current;

  const toggleMenu = () => {
    Animated.timing(animation, {
      toValue: expanded ? 0 : 1,
      duration: 300,
      useNativeDriver: false,
    }).start();
    setExpanded(!expanded);
    onToggleMenu();
  };

  const heightInterpolate = animation.interpolate({
    inputRange: [0, 1],
    outputRange: [0, 50],
  });

  const opacityInterpolate = animation.interpolate({
    inputRange: [0, 1],
    outputRange: [0, 1],
  });

  const toggleSwitch = React.useCallback(() => {
    setIsEnabled(!isEnabled);
    onAction('FAVORITES', !isEnabled);
  }, [isEnabled, onAction]);

  return (
    <>
      <View style={styles.container}>
        <View style={styles.container}>
          <Text style={styles.switchText}>
            {isEnabled ? 'Ocultar favoritos' : 'Ver favoritos'}
          </Text>
          <Switch
            trackColor={{false: '#767577', true: 'rgb(0, 240, 104)'}}
            thumbColor={'#f4f3f4'}
            ios_backgroundColor="#3e3e3e"
            onValueChange={toggleSwitch}
            value={isEnabled}
            style={{transform: [{scaleX: 0.8}, {scaleY: 0.8}]}}
            testID="toggle-switch"
          />
        </View>
        <TouchableOpacity
          onPress={toggleMenu}
          style={styles.toggleButton}
          testID="toggle-button">
          <SearchIcon />
        </TouchableOpacity>
      </View>
      <Animated.View
        style={[
          styles.menuContainer,
          {height: heightInterpolate, opacity: opacityInterpolate},
        ]}
        testID="menu-container">
        <InputSearch
          onCancel={toggleMenu}
          onChangeText={value => onAction('SEARCH', value)}
        />
      </Animated.View>
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignSelf: 'flex-end',
    alignItems: 'center',
  },
  toggleButton: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    paddingLeft: 10,
  },
  menuContainer: {
    overflow: 'hidden',
    marginTop: 10,
  },
  switchText: {
    color: 'white',
    fontSize: 12,
  },
});
