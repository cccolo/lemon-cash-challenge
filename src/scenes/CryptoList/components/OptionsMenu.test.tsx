import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import {OptionsMenu} from './OptionsMenu';

describe('OptionsMenu Component', () => {
  it('matches snapshot', () => {
    const {toJSON} = render(
      <OptionsMenu onToggleMenu={jest.fn} onAction={jest.fn} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('toggles switch and calls onAction', () => {
    const onActionMock = jest.fn();
    const {getByTestId} = render(
      <OptionsMenu onAction={onActionMock} onToggleMenu={jest.fn} />,
    );

    const switchElement = getByTestId('toggle-switch');
    fireEvent(switchElement, 'valueChange', true);
    expect(onActionMock).toHaveBeenCalledWith('FAVORITES', true);
  });
});
