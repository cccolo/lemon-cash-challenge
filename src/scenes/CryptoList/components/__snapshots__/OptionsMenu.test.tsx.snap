// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsMenu Component matches snapshot 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "alignSelf": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "alignSelf": "flex-end",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 12,
          }
        }
      >
        Ver favoritos
      </Text>
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="rgb(0, 240, 104)"
        style={
          [
            {
              "height": 31,
              "width": 51,
            },
            [
              {
                "transform": [
                  {
                    "scaleX": 0.8,
                  },
                  {
                    "scaleY": 0.8,
                  },
                ],
              },
              {
                "backgroundColor": "#3e3e3e",
                "borderRadius": 16,
              },
            ],
          ]
        }
        testID="toggle-switch"
        thumbTintColor="#f4f3f4"
        tintColor="#767577"
        value={false}
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "opacity": 1,
          "paddingLeft": 10,
        }
      }
      testID="toggle-button"
    >
      <svgMock />
    </View>
  </View>,
  <View
    collapsable={false}
    style={
      {
        "height": 0,
        "marginTop": 10,
        "opacity": 0,
        "overflow": "hidden",
      }
    }
    testID="menu-container"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderColor": "white",
            "borderRadius": 5,
            "borderWidth": 1,
            "color": "white",
            "flex": 1,
            "flexDirection": "row",
            "height": 40,
            "marginBottom": 10,
            "paddingLeft": 10,
          }
        }
      >
        <svgMock />
        <TextInput
          onChangeText={[Function]}
          placeholder="Ingresa nombre o simbolo"
          placeholderTextColor="white"
          style={
            {
              "color": "white",
              "maxWidth": 260,
              "paddingLeft": 10,
            }
          }
          testID="input"
          value=""
        />
        <View
          style={
            {
              "alignItems": "flex-end",
              "flex": 1,
              "marginRight": 12,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="cancel-button"
          >
            <svgMock
              style={
                {
                  "color": "white",
                  "textAlign": "right",
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>,
]
`;
